### PR TITLE
Walking over your friends with murderous intent or while fat will now trample them [WIP]

### PR DESCRIPTION
**Pull request by Citinited**
This PR implements trampling!

Only mobs that are human may trample or be trampled (meaning drones, pets etc. are safe). Mobs may only be trampled if they are lying on the ground, and the damage dealt to them is dependent on multiple factors, such as the trampling mob's footwear.

I've also added an `is_fat()` proc to `mob/living/carbon/human/human.dm` to check if a human is either obese from genetics or merely from overeating. (Relevant to other PRs!) 

:cl:
add: Human mobs can now get trampled by other mobs if they are on the floor. The attacker must have a non-help intent, or be fat. Certain shoes are also better at trampling mobs than other shoes.
add: Added mantreads and hobnails, Mantreads deal ridiculous trampling damage to prone mobs. Hobnails are craftable items which increase shoes' attacking capabilities at the cost of slowdown and being loud.
/:cl:

EDIT: quick numbers, for anyone who wants to know roughly how much damage this will do.
Under normal circumstances you will do 3 damage per stomp.
Being fat will bump this up to 4 damage.
Wearing magboots will deal 3.75 damage.
Being fat and wearing magboots will deal roughly 5 damage.
You can trample at most every 1.5 seconds - you can't trample someone a hundred times with one stun.

### CODE REVIEW

**Comment by alexkar598** [comment]
Why would this be good for the game?

**Comment by Spacemanspark** [comment]
As amusing as this would be, this severely buffs the telebaton.

**Comment by Citinited** [comment]
The thought was that you would have to watch where you're standing, for example you'd need to make sure you weren't accidentally stamping on people in medbay, escape, or just randomly lying in the corridors.

People with telebatons / stunprods can still punch mobs they stun, and it's not all that much damage.

I can readjust the numbers / come up with some way to implement it so you can't really trample someone as a viable method of injuring them (i.e. it's a crowd that has to trample, not just an individual).

**Comment by SamHPurp** [comment]
So I could simply slip someone, then walk back and forth over them over and over, killing them before they can stand?

**Comment by Citinited** [comment]
@SamHPurp No, there is a cooldown of 1.5 seconds between each trample.

**Comment by scrubmcnoob** [comment]
Given how many people are not always on help intent, this will create a ton of situations of someone getting tased, then getting trampled over by another person trying to help them.

**Comment by Citinited** [comment]
If I added a thing to the disarm intent so you couldn't disarm + trample immediately, would that be good?

**Comment by Tayyyyyyy** [comment]
This is a buff to stun/slip based combat. It increases the DPS you can deal to someone on the ground, which is where I see potential issues with balance. I like the idea that you can't just randomly stand on people anymore, though. 

What would happen in scenarios where like 10 people squeeze into an escape pod lying down? Would you be able to damage all those people at once? I guess that might be an intentional thing if the idea is to prevent people from lying down in unrealistic scenarios.

**Comment by Citinited** [comment]
@Tayyyyyyy That could be done via a proc to check the number of mobs on a specific tile, and do a thing if there are more than X mobs on it. (That'd probably be a separate PR, though). As it stands with this PR, if you stamp on some mobs, only one will actually be damaged.

@scrubmcnoob Indeed. Those people would need to watch their step! 😄

**Comment by alexkar598** [comment]
And sad people? They will come back with the need to go to medbay

**Comment by alexkar598** [comment]
Meant SSd people.damm auto correct

**Comment by Citinited** [comment]
Oh, shit, SSDs 🤔

**Comment by granodd** [comment]
_Forget to turn off harm intent and step over an SSD_

DING DONG BANNU!

Not sure how to feel about this. It'd be neat, but at the same time, it seems like it'd cause a lot of trouble.

**Comment by alexkar598** [comment]
How would it be explained to new players? It's not a feature that other code bases that I know of have this so it's making a complicated game even more complicated.

**Comment by Citinited** [comment]
It wouldn't, but since you get a big message after you do it, you'd probably be able to figure it out. (And yes I know newbies do randomly punch people for no reason, but this isn't all that different to learning that "hitting someone with a crowbar injures them")

**Comment by tigercat2000** [comment]
/tg/ tried this, and reverted it- You really shouldn't be able to unintentionally cause damage to people.

**Comment by Citinited** [comment]
Interesting -- before I coded it I looked at /tg/ to see if they had done it but couldn't find it on google (admittedly, I didn't look too deeply).

I can think of a couple ways this PR can go, but will wait for further feedback before enacting anything.

**Comment by KasparoVy** [comment]
Stun and walk over people as fast as you can with methboots.

**Comment by Fox-McCloud** [comment]
>/tg/ tried this, and reverted it- You really shouldn't be able to unintentionally cause damage to people.

Yeah, they did, and a vocal minority repeatedly brought up how bad of an idea it would be and the amount of accidental deaths/crits it would cause, not to mention the plausible deniability as a griefing tool is astoundingly high.

They went through with it, and it was reverted in less than 2 weeks; it was just as bad, if not worse, than the individuals suggested.

**Comment by taukausanake** [comment]
I'd argue, like any feature ever implemented, you'd have to get used to how it works whenever it's good or bad. I think this is nice and more immersive; pretty funny to walk over someone not giving a fuck. It shouldn't deal that much damage and the cooldown is good. I'd like to see an implementation where if someone was running, on any intent, that they have a chance to trip over someone laying down. Give tactical resting an actual use rather than wavedashing and bending over to trip can be a good tactic.

I'd like to see this in the server for a little. Be interesting to see how it goes with us

**Comment by coldud13** [comment]
This sounds painful if many people cram into an escape pod.

**Comment by taukausanake** [comment]
@coldud13 as it should. Escape pods aren't clown cars after all

**Comment by deep-fried-faceplant** [comment]
Make jackboots deal extra damage.

**Comment by Citinited** [comment]
Not sure if this will fly in its current form, so there's a couple ways I can go from here.

1. I change the behaviour of trampling so that you trip over people lying on the floor (maybe taking some brute) instead of dealing damage to the prone mob. This would deal with any issues relating to griefing and damaging SSDs.

2. I keep the behaviour as-is, but restrict it to a few (rare) pieces of footwear. Perhaps a traitor item could include this?

3. I modify trampling so that only crowds can stampede over people (instead of one person being able to stomp on people). This does not fully address the issue with SSDs.

As for hobnails - I think those could be kept, as if you stand on people with them, your intentions are immediately obvious.

**Comment by SamHPurp** [comment]
Could simply increase their damage when they un-arm'd hit someone that is lying down.

**Comment by Vivalas** [comment]
The intention is to make people more cautious of walking over people though.

**Comment by ZomgPonies** [comment]
Lower the base damage value to 1, and keep the other buffs and it should be fine. Think of it as stepping on people's hands. Make it such minor damage that it can't really be used as an effective combat buff while still causing some damage if there's trampling to get on the escape shuttle for example. 

tl;dr: Make it more thematic than actually deadly.

**Comment by alexkar598** [comment]
Can Borgs trample people?

**Comment by taukausanake** [comment]
I'd argue borgs shouldn't trample people. I don't know why, just rather they didn't. Unless they have treads. That would be funny

**Comment by Tayyyyyyy** [comment]
To answer your question, no, at the moment the PR only supports humans trampling.

**Comment by taukausanake** [comment]
Does this work when someone is walking or running or both? I think if someone is taking the time to walk they won't trample someone. Maybe step on them with a gut shot (not really) but not trample. Walking would show they are aware there is someone on the ground and they are being careful around them.

You think in the future it would be possible to stand on someone to keep them from standing up? Just a side thought.

Tripping over SSD people is genius :+1:

**Comment by Citinited** [comment]
@taukausanake Actually, that's not a bad idea (and the reasoning is similar to that of help intent not trampling the mob). I'll change that now 👍 

As for standing on them to keep them standing up, I know pinning has already been suggested and was, ultimately, rejected. (See #981) This would be very similar to that, so it most likely wouldn't go through.

File: code/modules/mob/living/carbon/human/human.dm [hunk 1]
```diff
@@ -539,13 +542,51 @@
...
+		var/obj/item/clothing/shoes/S = shoes
+		if(S.trampling_coefficient == 0) //No point treading on them if we're going to do 0 damage
+			return
+		D *= S.trampling_coefficient
+	else
+		D *= 0.25 //Standing on someone without shoes isn't that damaging
+	if(H.stat == DEAD)
+		H.visible_message("<span class = 'warning'>[src] treads upon [H]'s lifeless form.</span>", "<span class = 'warning'>[src] further violates your corpse.</span>", "<span class = 'warning'>You hear a fleshy thud.</span>")
+	else
+		H.visible_message("<span class = 'danger'>[src] stomps on [H]!</span>", "<span class = 'userdanger'>[src] squashes you under their feet!</span>", "<span class = 'warning'>You hear a fleshy thud.</span>")
```
Comment (Citinited) [inline code/modules/mob/living/carbon/human/human.dm]: I'm sure there's a way to combine this `if... else...` into one line. Not sure though 🤔

Comment (taukausanake) [inline code/modules/mob/living/carbon/human/human.dm]: Like an inline if statement? Wouldn't that be more ugly and long?

Comment (Citinited) [inline code/modules/mob/living/carbon/human/human.dm]: No, I mean something like `H.visible_message("You were trampled!" || if H.stat == "DEAD" "Your corpse was trampled!")`

**Comment by taukausanake** [comment]
> See 981

HOW OLD ARE YOU?

**Comment by Citinited** [comment]
I've been imprisoned here since February 2015 ;P

**Comment by Fox-McCloud** [comment]
Going to turn this one down.

The ability to unintentionally hurt someone purely by a passive action will be annoying at best, and incredibly harmful, at worst. This has been tried elsewhere to immense detriment and frustration. On a codebase like ours, where brute damage is even stronger, this is likely to only create more problems.

Likewise, the plausible deniability of using this for grief is incredibly high.

**Comment by Citinited** [comment]
Should I assume that the idea of trampling is completely unworkable? I'd left it as [WIP] so I could address the damage thing...

**Comment by Fox-McCloud** [comment]
Yeah, I don't think it's a good idea.

**Comment by Citinited** [comment]
*shrug* Fair enough.
